### PR TITLE
Improve handling Esc during message compose

### LINKED ISSUE #1342: Improve handling of pressing `Esc` during message compose
Currently `Esc` can be pressed to exit compose, without sending the message.

This can be useful if one wants to cancel the send, but can also be frustrating if accidentally hitting the key after having written a good message which you intend to send.

This can be made worse, if a key is near the `Esc` key and used for composition, eg. the <kbd>`</kbd> key for inline code or quote/code blocks.

Ideas for resolving this:
- Prompt with a confirmation popup to exit compose, if there is more than some threshold of content present (could be empty message, could be 5, 10, 15, ... characters)
- Prompt with a variation on a confirmation popup, giving the option to save as a draft
- Automatically save as a draft, but prompt if a draft already exists (like we do now)
- Automatically save as a draft - *only possible if we support more than one draft in the session, ie. improve draft handling generally ~ #302)

Please discuss this in the stream before proceeding with this, since it would be good to get more views on this first.

### PR DESCRIPTION
**Pull request by rsashank**
### What does this PR do, and why?
Uses the existing confirmation popup to confirm whether user wants to exit the compose box upon pressing `Esc`, Enabling the user to go back and save their draft from compose.

### Outstanding aspect(s)    

- [ ] 

### External discussion & connections

- [x] Discussed in **#zulip-terminal** in `Improve handling Esc during msg compose #T1342 #T1362 #T1442`
- [x] Fully fixes #1342 
- [ ] Partially fixes issue #
- [x] Builds upon previous unmerged work in PR #1362 
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?

- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [ ] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [ ] It has a commit summary describing the  motivation and reasoning for the change
- [ ] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### CODE REVIEW

File: zulipterminal/ui_tools/boxes.py [hunk 1]
```diff
@@ -801,6 +807,8 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
                     )
         elif is_command_key("GO_BACK", key):
             self.send_stop_typing_status()
+            if len(self.msg_write_box.edit_text) > 10:
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py:810]: Generally these literal values (ie. like 10) are good to extract as named constants. That makes it clearer what the value is at the point where it is used, and also means that one doesn't necessarily need to search through the code to determine what a value/threshold is.

File: zulipterminal/core.py [hunk 2]
```diff
@@ -523,6 +523,22 @@ def stream_muting_confirmation_popup(
         mute_this_stream = partial(self.model.toggle_stream_muted_status, stream_id)
         self.loop.widget = PopUpConfirmationView(self, question, mute_this_stream)
 
+    def exit_compose_confirmation_popup(self) -> None:
+        question = urwid.Text(
+            (
+                "bold",
+                "Please confirm that you wish to exit the compose box. "
+                "(You can save the message as a draft upon returning to compose)",
```
Comment (neiljp) [inline zulipterminal/core.py:531]: The bottom line wraps partially for me to be on the same line as the first - I think it makes sense to have these at least on separate lines.

**Review by neiljp — COMMENTED** [review]
@rsashank Thanks for working on this :+1:

This does appear to work well, though the control flow looks a little confusing - see my inline comment.

Re the tests, there is only one failing, though it would be useful to add a test for the new behavior too.

To understand the tests generally , I'd suggest ensuring you at a minimum understand `fixtures` (eg. `write_box` in the failing test), and how `parametrize` works - see the pytest documentation, or read our test code of course :) Also knowing how `Mock`s work (often accessed via `mocker` in pytest) would be useful.

For the specific issue you see failing, see the failing line - it thinks the object doesn't have the appropriate attribute. There are times when we need to specifically set a value, but I suspect the issue here is that this test doesn't initialize a particular view (DM, DM-edit, stream, stream-edit), so the box(es) don't get initialized, including the one you are trying to access. For now I would suggest setting the value on the write_box in the test to a specific value, which may be easier using a `Mock`.

File: zulipterminal/ui_tools/boxes.py [hunk 3]
```diff
@@ -801,6 +807,8 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
...
+                self.view.controller.exit_compose_confirmation_popup()
             self._set_compose_attributes_to_defaults()
```
[the remark this answers is not in the document]
Comment (rsashank) [reply zulipterminal/ui_tools/boxes.py]: I believe the popup doesn't drag the control flow from this method. However, it seems that `self.loop.run()` plays a role in preserving the WriteBox, as removing that line causes the WriteBox to disappear. This suggests that the popup is not dragging control flow out of the method.

However, I have modified the logic by incorporating an if-else statement and re-using the `exit_compose_box` method in the else statement. Also the new logic, excludes the need for the `self.loop.run()` line. 

I'm working on the tests. Will keep this updated.

**Comment by rsashank** [comment]
I've updated the tests. The initial functionality of immediately exiting the message compose now only applies when the character count is below 10. To accommodate this, I modified the `write_box.msg_write_box.edit_text` from `"random text"` to `"random"`. Additionally set the value on the write_box to `""` using `Mock` to fix the `AttributeError`, as suggested earlier.

It now passes all the existing tests.

I am currently working on modifying the tests to account for the scenario where there are more than 10 characters.

**Comment by rsashank** [comment]
@zulipbot add PR needs review

**Comment by rsashank** [comment]
@zulipbot add "PR needs review"

**Comment by rsashank** [comment]
@zulipbot remove "PR awaiting update"

**Comment by rsashank** [comment]
Included two additional tests to verify the popup being triggered for messages exceeding 10 characters (adjustable with the `confirmation_msg_length` variable).

**Comment by rsashank** [comment]
Updated this PR. Currently it uses 15 character length as the threshold to determine whether to trigger the popup or not.

File: tests/ui_tools/test_boxes.py [hunk 4]
```diff
@@ -235,15 +235,39 @@ def test__compose_attributes_reset_for_private_compose(
...
         size = widget_size(write_box)
         write_box.keypress(size, key)
 
+        write_box.view.controller.exit_compose_confirmation_popup.assert_not_called()
         assert write_box.to_write_box is None
         assert write_box.msg_write_box.edit_text == ""
         assert write_box.compose_box_status == "closed"
 
+    @pytest.mark.parametrize("key", keys_for_command("GO_BACK"))
+    def test__compose_attributes_reset_for_private_compose_popup(
```
Comment (neiljp) [inline tests/ui_tools/test_boxes.py:251]: It is useful to separate the function name from the outcome/action by an extra underscore, so here before the `popup`, so it doesn't read as part of the function name.

It would also be useful to rename the existing related tests to indicate they are no longer a 'plain' test but for a different expected outcome/action.

**Review by neiljp — COMMENTED** [review]
@rsashank Thanks for the update - this looks clearer :+1:

I left some comments inline, but they relate to the same point.

It would be useful to know the threshold below which the web app discards drafts, which would likely be a good equivalent threshold to use here, if it's different to 15?

I'd need to examine the code again, but wonder if we should adjust the existing tests to only assert on the new function being called, and have a separate test for the new function which defines that behavior. Then changes to the internals which that new function may need to address won't need to be updated in these tests. If so, that would be a good refactor commit.

File: tests/ui_tools/test_boxes.py [hunk 5]
```diff
@@ -254,15 +278,37 @@ def test__compose_attributes_reset_for_stream_compose(
...
+        key: str,
+        mocker: MockerFixture,
+        write_box: WriteBox,
+        widget_size: Callable[[Widget], urwid_Size],
+    ) -> None:
+        mocker.patch(WRITEBOX + "._set_stream_write_box_style")
+        write_box.stream_box_view(stream_id=1)
+
+        # Testing input more than 15 characters; >15 prompts confirmation popup
+        write_box.msg_write_box.edit_text = "random text (>15 characters)"
```
[the remark this answers is not in the document]
Comment (rsashank) [reply tests/ui_tools/test_boxes.py]: I've revised this to use a module-level constant. Using the `"." * CONFIRMATION_MSG_LENGTH*` ensures there are sufficient characters in the compose box for testing whether the popup is triggered. 

Similarly `"." * (CONFIRMATION_MSG_LENGTH - 1)*`  ensures there are less characters than the threshold in the compose box when testing no popup being triggered. 

Is this approach acceptable?

File: tests/ui_tools/test_boxes.py (hunk 4 above)
Comment (rsashank) [reply tests/ui_tools/test_boxes.py]: Revised the function names; are they appropriate now?

**Comment by rsashank** [comment]
@neiljp Thanks for the review :)

> I'd need to examine the code again, but wonder if we should adjust the existing tests to only assert on the new function being called, and have a separate test for the new function which defines that behavior. Then changes to the internals which that new function may need to address won't need to be updated in these tests. If so, that would be a good refactor commit.

Interesting, so we would only need to check for whether `exit_compose_confirmation_popup` is called? And then make another test for the behavior when that function is called? But, what about the case where the popup isn't activated? Shouldn't we retain the current assertions to cover that case as well?
```
assert write_box.to_write_box is None
assert write_box.msg_write_box.edit_text == ""
assert write_box.compose_box_status == "closed"
```

**Comment by rsashank** [comment]
@zulipbot remove "PR awaiting update"
@zulipbot add "PR needs review"

File: zulipterminal/ui_tools/boxes.py [hunk 6]
```diff
@@ -54,6 +54,9 @@
 from zulipterminal.urwid_types import urwid_Size
 
 
+CONFIRMATION_MSG_LENGTH = 15
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py:57]: This constant makes it more readable :+1:

Let's leave the value as it is for now - as a constant, it should be easy to adjust later if we need to.

We should probably do a sweep for this with other constants, but marking this as `Final` may help with accidental changing of the value later on.

I suspect we could improve the name here, avoiding the abbreviation, and indicating if this is a minimum/maximum. Expressive names are better :)

**Review by neiljp — COMMENTED** [review]
@rsashank Thanks for your last update - the tests look broadly good, and reviewing the test changes I suggested previously, that's fine to leave for now, though you could treat extracting `exit_compose_box` as a refactor and add a test for it in a prep commit. The idea for that was as you suspected I think, and we'd add a test for `exit_compose_box` to cover the expected behavior from that; the other asserts would then duplicate the new test, so could be removed.

I left a few inline comments. One of those relates to how the popup should interact with autocomplete, which we didn't think of previously.

Another slightly confusing aspect of the current implementation is that if one resumes compose after the popup and saves it as a draft, one is still prompted with the popup again. We did discuss how drafts might interact with this in the topic in #zulip-terminal; I'd suggest if the draft effectively matches the message then skipping the popup, and possibly showing a notice in the footer?

File: zulipterminal/ui_tools/boxes.py (hunk 6 above)
Comment (rsashank) [reply zulipterminal/ui_tools/boxes.py]: The variable has been updated to `MAX_MESSAGE_LENGTH_CONFIRMATION_POPUP`, indicating the maximum message length for which confirmation popup is not prompted. Should I add a comment for further clarity, or is the name itself clear enough?

Or should I change it to `MAX_MESSAGE_LENGTH_NO_CONFIRMATION_POPUP`

File: tests/ui_tools/test_boxes.py [hunk 7]
```diff
@@ -1508,13 +1555,15 @@ def test_keypress_SEND_MESSAGE_no_topic(
...
         self,
+        mocker: MockerFixture,
         write_box: WriteBox,
         widget_size: Callable[[Widget], urwid_Size],
         current_typeahead_mode: bool,
         expected_typeahead_mode: bool,
         expect_footer_was_reset: bool,
         key: str,
     ) -> None:
+        write_box.msg_write_box = mocker.Mock(edit_text="")
```
[the remark this answers is not in the document]
Comment (rsashank) [reply tests/ui_tools/test_boxes.py:1572]: That sounds logical 👍

I've relocated the logic for exiting the typeahead to the new `exit_compose_box` method, which works for this.

**Comment by rsashank** [comment]
Updated this PR :)

@zulipbot remove "PR awaiting update"
@zulipbot add "PR needs review"

**Comment by Niloth-p** [comment]
This implementation LGTM.
Those are some interesting edge cases indeed.
I've got to go apply that previous feedback about the splitting of tests to several of my commits :)

**Review by neiljp — COMMENTED** [review]
@rsashank I just caught up with the channel topic regarding this, so was able to complete a review I started previously :)

To clarify some points that were raised here and there:
- There's no need to consider auto-saving at this point; let's get this feature in, and then we could consider how any drafts feature that is implemented affects this later
- Regarding prompting on editing messages, it seems like the web app requires an explicit 'cancel' vs 'save' to do either action. So, we *could* have a confirmation popup for discarding 'drafts' of edits to messages, with a different message. That said, edits are *probably* more likely to be smaller corrections, with less likelihood of data-loss, so that could be a followup.
- I think the draft confirmation being centered may be clearer, as long as it's very obviously distinct from eg. the 'discard compose' confirmation. I looked into refactoring those popups, and I think a design along those lines may be useful, eg. with styling based on areas/tasks like the other popups.

File: tests/ui_tools/test_boxes.py [hunk 8]
```diff
@@ -1582,10 +1576,33 @@ def test_keypress_typeahead_mode_autocomplete_key(
...
+
+    @pytest.mark.parametrize(
+        "key, current_typeahead_mode, expected_typeahead_mode",
+        [
+            (primary_key_for_command("GO_BACK"), True, False),
+            ("space", True, False),
+            ("k", True, False),
+        ],
+    )
+    def test_keypress_typeahead_mode_autocomplete_key_footer_reset(
```
[the remark this answers is not in the document]
Comment (rsashank) [reply tests/ui_tools/test_boxes.py]: What's the simplification that can be done in terms of parametrize values? I'm a bit confused 😅.

Oh true, maybe we can should separate GO_BACK into its own test. Since reset actions are dependent on the length of the message in compose. What do you think?

File: zulipterminal/core.py [hunk 9]
```diff
@@ -510,10 +510,17 @@ def search_messages(self, text: str) -> None:
 
     def save_draft_confirmation_popup(self, draft: Composition) -> None:
         question = urwid.Text(
-            "Save this message as a draft? (This will overwrite the existing draft.)"
+            (
+                "bold",
+                "Save this message as a draft? "
+                "(This will overwrite the existing draft.)",
+            ),
+            "center",
```
[the remark this answers is not in the document]
Comment (rsashank) [reply zulipterminal/core.py:523]: It seems that centering the popup doesn't cover the compose box text, even on smaller terminal sizes, except for extreme cases where the middle column is barely visible (What is the minimum terminal size we look to support?).

However, centering the popup does enhance the visual appeal, though that's subjective. I can drop this commit if needed :)

File: zulipterminal/ui_tools/boxes.py [hunk 10]
```diff
@@ -807,12 +815,23 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
...
+            )
+            not_saved_as_draft = (
+                saved_draft is None
+                or self.msg_write_box.edit_text != saved_draft.get("content")
+            )
+
+            if compose_not_in_edit_mode and exceeds_max_length and not_saved_as_draft:
+                self.view.controller.exit_compose_confirmation_popup()
+            else:
+                self.exit_compose_box()
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py:834]: This logic is a lot clearer :+1: I hope you agree :)

Nit: consider a variable for `...edit_text`, perhaps instead of the message length?

Nit: group the initial/middle data setup (saved_draft, compose_max_length/equivalent above), then the derived bools - or each prep+bool together, if not reused.

File: zulipterminal/ui_tools/boxes.py [hunk 11]
```diff
@@ -710,6 +715,7 @@ def autocomplete_emojis(
         return emoji_typeahead, emojis
 
     def exit_compose_box(self) -> None:
+        self._set_default_footer_after_autocomplete()
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py:718]: Having extracted this function in the earlier commit, it would be useful to confirm why we need it here in the commit body.

File: zulipterminal/core.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/core.py:523]: See the FAQ - ideally we're aiming to still support 80x25.

Keeping things in separate commits works just fine, as you say, though if content is not visible we could bring it into the popup or similar, as I think we discussed. So certainly no need to drop in advance :)

BTW the other reason for keeping commits separate is that if we later decide to adjust it, we can potentially revert a single commit, or use it as the basis for a change to a previous behavior.

File: tests/ui_tools/test_boxes.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply tests/ui_tools/test_boxes.py]: Re simplification: the keys vary, but the modes before and after apparently do not :)

Re my second comment, it likely needs covering, though may be worth being careful that it isn't over-tested, ie. in multiple places.

**Comment by rsashank** [comment]
@neiljp @mounilKshah, does this require any additional changes? I have self-reviewed the code again and @Niloth-p had buddy-reviewed this before. I can't seem to find anything that still needs to be addressed, apart from handling the exit when in editing mode (which I will address in another PR).

File: tests/ui_tools/test_boxes.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply tests/ui_tools/test_boxes.py]: The simplification is how the modes are the same in every case now, in each of these new tests.

**Comment by neiljp** [comment]
@rsashank Other than resolving the expected conflicts you expected with the other PR, this looks essentially good to go except for a few tiny nits from previous comments :+1:
(that other PR looks like it'll help the readability of this code too)

I did notice when I was manually testing to confirm the autocomplete aspect we discussed, that the recovery of the autocomplete after cancelling exit of compose doesn't appear to quite work properly - the footer is present, and we're presumably still 'in autocomplete mode' via that bool, but it has '[No matches found]'? I'm not sure how easy that is to resolve at this point?

If you can resolve that fairly easily, that'd be great, but if not then I'd suggest adding a call to that helper function with a comment regarding this problem, so we can simply reset it for now but it's noted in the code for future reference.

It'll be useful to have this merged - I lost a composition just recently...!

**Comment by rsashank** [comment]
> @rsashank Other than resolving the expected conflicts you expected with the other PR, this looks essentially good to go except for a few tiny nits from previous comments 👍 (that other PR looks like it'll help the readability of this code too)
> 
> I did notice when I was manually testing to confirm the autocomplete aspect we discussed, that the recovery of the autocomplete after cancelling exit of compose doesn't appear to quite work properly - the footer is present, and we're presumably still 'in autocomplete mode' via that bool, but it has '[No matches found]'? I'm not sure how easy that is to resolve at this point?
> 
> If you can resolve that fairly easily, that'd be great, but if not then I'd suggest adding a call to that helper function with a comment regarding this problem, so we can simply reset it for now but it's noted in the code for future reference.
> 
> It'll be useful to have this merged - I lost a composition just recently...!

In that case, could we maybe have this merged? I'll look into this once I tidy up the polls and spoilers PRs and address it separately.

**Comment by neiljp** [comment]
@rsashank Thanks for keeping this work going - this is going to save many lost compositions, I suspect! 👍 

I just pushed back here with a removal of the `EXIT_COMPOSE` branch near the top of the `keypress()` method, and a corresponding comment with TODO there, matching my last note about autocomplete not yet resuming correctly.

You noted the early push I made had a test issue - that was due to the footer being reset twice in some cases, which I don't think will be an issue, but I've added a comment in any case, as well as migrating that change across the commits.

I did want to get the follow-up points noted, but I'm going to merge this first :tada:
